chainimport/headers_import.go

```diff
@@ -0,0 +1,1814 @@
+	}
+
+	// Get and verify filter headers. We can safely use the same import
+	// source index used for import source block header store since we've
+	// validated earlier their start height and headers count match exactly.
+	importSourceHeader, err = s.FilterHeadersImportSource.GetHeader(
+		importSourceIndex,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to get filter header from import "+
+			"source at height %d: %w", height, err)
+	}
+
+	// Type assert the header retrieved from import source to be filter
+	// header.
+	importSourceFilterHeader, ok := importSourceHeader.(*FilterHeader)
+	if !ok {
+		return fmt.Errorf("expected FilterHeader type, got %T",
+			importSourceHeader)
+	}
+
+	// Get the equivalent filter header from target store.s
+	filterHeaderStore := s.options.TargetFilterHeaderStore
+	targetFilterHeader, err := filterHeaderStore.FetchHeaderByHeight(
+		height,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to get filter header from target at "+
+			"height %d: %w", height, err)
+	}
+
+	// Compare filter headers from import and target sources.
+	sourceFilterHeaderHash := importSourceFilterHeader.FilterHash
+	targetFilterHeaderHash := *targetFilterHeader
+	if !sourceFilterHeaderHash.IsEqual(&targetFilterHeaderHash) {
+		return fmt.Errorf("filter header mismatch at height %d: "+
+			"source has %v but target has %v", height,
+			sourceFilterHeaderHash, targetFilterHeaderHash)
+	}
+
+	log.Debugf("Headers from %s (block) and %s (filter) verified at "+
+		"height %d", s.BlockHeadersImportSource.GetURI(),
+		s.FilterHeadersImportSource.GetURI(), height)
+
+	return nil
+}
+
+// processNewHeadersRegion imports headers from the specified region into the
+// target stores. This method handles the case where headers exist in the import
+// source but not in the target stores.
+func (s *HeadersImport) processNewHeadersRegion(region HeaderRegion,
+	result *ImportResult) error {
+
+	if !region.Exists {
+		return nil
+	}
+
+	log.Infof("Adding %d new headers (block and filter) from heights "+
+		"%d to %d", region.End-region.Start+1, region.Start, region.End)
+
+	err := s.appendNewHeaders(region.Start, region.End)
+	if err != nil {
+		return fmt.Errorf("failed to append new headers: %w", err)
+	}
+
+	result.ProcessedCount += int(region.End - region.Start + 1)
+	result.AddedCount += int(region.End - region.Start + 1)
+
+	return nil
+}
+
+// appendNewHeaders adds new headers from import source.
+func (s *HeadersImport) appendNewHeaders(startHeight, endHeight uint32) error {
+	// Get metadata from sources.
+	metadata, err := s.BlockHeadersImportSource.GetHeaderMetadata()
+	if err != nil {
+		return fmt.Errorf("failed to get block header "+
+			"metadata: %w", err)
+	}
+
+	totalHeaders := endHeight - startHeight + 1
+	log.Infof("Appending %d new headers in batches of %d", totalHeaders,
+		s.options.WriteBatchSizePerRegion)
+
+	// Process headers in batches.
+	for batchStart := startHeight; batchStart <= endHeight; {
```

Comment:
I suspect this whole loop might be where we could potentially squeeze some performance in terms of memory allocations out of the code. So it would be great if we had an actual Golang `BenchmarkXXX` test that covers this, so we can then extract memory and CPU profiles from it and optimize.
But not blocking for now, can be done in a future PR.

Reply:
Ah, yeah, I forgot about that test. Sounds good to me.